File: alle-core/src/main/java/io/github/shomah4a/alle/core/keybind/KeyStroke.java
package io.github.shomah4a.alle.core.keybind;

import java.util.Objects;
import org.eclipse.collections.api.factory.Sets;
import org.eclipse.collections.api.set.ImmutableSet;

/**
 * 単一のキー入力。修飾キーの組み合わせとキーコードで構成される。
 * 例: Ctrl+X, Meta+F など。
 */
public record KeyStroke(ImmutableSet<Modifier> modifiers, int keyCode) {

    // 特殊キー用コードポイント（Unicode Private Use Area）
    public static final int ARROW_UP = 0xF700;
    public static final int ARROW_DOWN = 0xF701;
    public static final int ARROW_LEFT = 0xF702;
    public static final int ARROW_RIGHT = 0xF703;

    /**
     * 修飾キーなしのキーストロークを生成する。
     */
    public static KeyStroke of(int keyCode) {
        return new KeyStroke(Sets.immutable.empty(), keyCode);
    }

    /**
     * 修飾キー付きのキーストロークを生成する。
     */
    public static KeyStroke of(int keyCode, Modifier... modifiers) {
        return new KeyStroke(Sets.immutable.with(modifiers), keyCode);
    }

    /**
     * Ctrl+キーのキーストロークを生成する。
     */
    public static KeyStroke ctrl(int keyCode) {
        return new KeyStroke(Sets.immutable.of(Modifier.CTRL), keyCode);
    }

    /**
     * Meta+キーのキーストロークを生成する。
     */
    public static KeyStroke meta(int keyCode) {
        return new KeyStroke(Sets.immutable.of(Modifier.META), keyCode);
    }

    /**
     * Shift+キーのキーストロークを生成する。
     */
    public static KeyStroke shift(int keyCode) {
        return new KeyStroke(Sets.immutable.of(Modifier.SHIFT), keyCode);
    }

    /**
     * キーストロークをEmacs風の表示文字列に変換する。
     * 例: Ctrl+x → "C-x", Meta+x → "M-x", 通常文字 → "a"
     */
    public String displayString() {
        // Shift+Tab は Emacs に合わせて <backtab> と表示する
        if (keyCode == '\t' && modifiers.equals(Sets.immutable.of(Modifier.SHIFT))) {
            return "<backtab>";
        }
        var sb = new StringBuilder();
        if (modifiers.contains(Modifier.CTRL)) {
            sb.append("C-");
        }
        if (modifiers.contains(Modifier.META)) {
            sb.append("M-");
        }
        if (modifiers.contains(Modifier.SHIFT)) {
            sb.append("S-");
        }
        sb.append(keyCodeToString(keyCode));
        return sb.toString();
    }

    private static String keyCodeToString(int keyCode) {
        return switch (keyCode) {
            case ARROW_UP -> "<up>";
            case ARROW_DOWN -> "<down>";
            case ARROW_LEFT -> "<left>";
            case ARROW_RIGHT -> "<right>";
            case '\n' -> "RET";
            case ' ' -> "SPC";
            case 0x7F -> "DEL";
            case '\t' -> "TAB";
            case 0x1B -> "ESC";
            default -> {
                if (Character.isValidCodePoint(keyCode) && !Character.isISOControl(keyCode)) {
                    yield String.valueOf(Character.toChars(keyCode));
                }
                yield String.format("<%d>", keyCode);
            }
        };
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof KeyStroke other)) return false;
        return keyCode == other.keyCode && modifiers.equals(other.modifiers);
    }

    @Override
    public int hashCode() {
        return Objects.hash(modifiers, keyCode);
    }
}

File: alle-script/src/main/resources/io/github/shomah4a/alle/script/alle/keystroke.py
"""KeyStroke の Python ラッパー。"""

from __future__ import annotations

from typing import Any

import java

_Modifier: Any = java.type('io.github.shomah4a.alle.core.keybind.Modifier')


class KeyStroke:
    """Java 側の KeyStroke をラップし、Python 的なインターフェースを提供する。

    :param java_keystroke: Java 側の KeyStroke インスタンス
    :type java_keystroke: Any
    """

    def __init__(self, java_keystroke: Any) -> None:
        self._ks: Any = java_keystroke

    @property
    def key_code(self) -> int:
        """キーコードを返す。

        :return: キーコード（Unicode コードポイント）
        :rtype: int
        """
        return self._ks.keyCode()

    @property
    def modifiers(self) -> frozenset[str]:
        """修飾キーの集合を返す。

        :return: 修飾キー名の集合（``"ctrl"``, ``"meta"``, ``"shift"``）
        :rtype: frozenset[str]
        """
        result: set[str] = set()
        java_modifiers = self._ks.modifiers()
        if java_modifiers.contains(_Modifier.CTRL):
            result.add("ctrl")
        if java_modifiers.contains(_Modifier.META):
            result.add("meta")
        if java_modifiers.contains(_Modifier.SHIFT):
            result.add("shift")
        return frozenset(result)

    def has_ctrl(self) -> bool:
        """Ctrl 修飾キーが含まれるかを返す。

        :return: Ctrl が含まれる場合 True
        :rtype: bool
        """
        return self._ks.modifiers().contains(_Modifier.CTRL)

    def has_meta(self) -> bool:
        """Meta 修飾キーが含まれるかを返す。

        :return: Meta が含まれる場合 True
        :rtype: bool
        """
        return self._ks.modifiers().contains(_Modifier.META)

    def has_shift(self) -> bool:
        """Shift 修飾キーが含まれるかを返す。

        :return: Shift が含まれる場合 True
        :rtype: bool
        """
        return self._ks.modifiers().contains(_Modifier.SHIFT)

    def display_string(self) -> str:
        """Emacs 風の表示文字列を返す。

        例: ``"C-x"``, ``"M-x"``, ``"a"``

        :return: 表示文字列
        :rtype: str
        """
        return self._ks.displayString()

    def __repr__(self) -> str:
        return f"KeyStroke({self.display_string()!r})"

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, KeyStroke):
            return NotImplemented
        return self._ks.equals(other._ks)

    def __hash__(self) -> int:
        return self._ks.hashCode()
